desc/optimize/_constraint_wrappers.py

```diff
@@ -0,0 +1,634 @@
+            Reduced state vector that satisfies linear constraints.
+
+        Returns
+        -------
+        g : ndarray
+            gradient vector.
+        """
+        x = self.recover(x_reduced)
+        df = self._objective.grad(x)
+        return df[self._unfixed_idx] @ self._Z
+
+    def hess(self, x_reduced):
+        """Compute Hessian of the sum of squares of residuals.
+
+        Parameters
+        ----------
+        x_reduced : ndarray
+            Reduced state vector that satisfies linear constraints.
+
+        Returns
+        -------
+        H : ndarray
+            Hessian matrix.
+        """
+        x = self.recover(x_reduced)
+        df = self._objective.hess(x)
+        return self._Z.T @ df[self._unfixed_idx, :][:, self._unfixed_idx] @ self._Z
+
+    def jac(self, x_reduced):
+        """Compute Jacobian of the vector objective function.
+
+        Parameters
+        ----------
+        x_reduced : ndarray
+            Reduced state vector that satisfies linear constraints.
+
+        Returns
+        -------
+        J : ndarray
+            Jacobian matrix.
+        """
+        x = self.recover(x_reduced)
+        df = self._objective.jac(x)
+        return df[:, self._unfixed_idx] @ self._Z
+
+
+class ProximalProjection(ObjectiveFunction):
+    """Remove equilibrium constraint by projecting onto constraint at each step.
+
+    Combines objective and equilibrium constraint into a single objective to then pass
+    to an unconstrained optimizer.
+
+    At each iteration, after a step is taken to reduce the objective, the equilibrium
+    is perturbed and re-solved to bring it back into force balance. This is analagous
+    to a proximal method where each iterate is projected back onto the feasible set.
+
+    Parameters
+    ----------
+    objective : ObjectiveFunction
+        Objective function to optimize.
+    constraint : ObjectiveFunction
+        Equilibrium constraint to enforce. Should be an ObjectiveFunction with one or
+        more of the following objectives: {ForceBalance, CurrentDensity,
+        RadialForceBalance, HelicalForceBalance}
+    eq : Equilibrium, optional
+        Equilibrium that will be optimized to satisfy the objectives.
+    verbose : int, optional
+        Level of output.
+    perturb_options, solve_options : dict
+        dictionary of arguments passed to Equilibrium.perturb and Equilibrium.solve
+        during the projection step.
+    """
+
+    def __init__(
+        self,
+        objective,
+        constraint,
+        eq=None,
+        verbose=1,
+        perturb_options={},
+        solve_options={},
+    ):
+        assert isinstance(objective, ObjectiveFunction), (
+            "objective should be instance of ObjectiveFunction." ""
+        )
+        assert isinstance(constraint, ObjectiveFunction), (
+            "constraint should be instance of ObjectiveFunction." ""
+        )
+        for con in constraint.objectives:
+            if not isinstance(
```

Comment:
The way its currently set up this is correct. The full set of constraints are parsed into linear and nonlinear in `_parse_constraints`, so this would only ever be called with the nonlinear equilibrium constraint

Reply:
OK so this class doesn't handle the linear constraints at all, and assumes that the `constraint` it is passed is just the nonlinear equilibrium constraint

Reply:
Yes, the linear constraints are handled by a different wrapper.